### PR TITLE
grpc-js-xds: Add support for k8s interop test framework

### PR DESCRIPTION
**Pull request by murgatroid99**
The test script is copied with slight modifications from [the Python test script](https://github.com/grpc/grpc/blob/master/tools/internal_ci/linux/grpc_xds_k8s_lb_python.sh). The docker image is large (a bit over 1GB), but when I tried to make it smaller with a build step like in [the Go Dockerfile](https://github.com/grpc/grpc-go/blob/master/interop/xds/client/Dockerfile), it caused the tests to fail and I couldn't tell why.

### CODE REVIEW

File: test/kokoro/xds_k8s_lb.cfg [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# Location of the continuous shell script in repository.
+build_file: "grpc-node/packages/grpc-js-xds/scripts/xds_k8s_lb.sh"
+timeout_mins: 180
+action {
+  define_artifacts {
+    regex: "artifacts/**/*sponge_log.xml"
+    regex: "artifacts/**/*sponge_log.log"
+    strip_prefix: "artifacts"
+  }
+}
```
Comment (sergiitk) [inline test/kokoro/xds_k8s_lb.cfg:26]: Nit: missing new line at the EOF.

File: packages/grpc-js-xds/scripts/xds_k8s_lb.sh [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  for test in "${test_suites[@]}"; do
+    run_test $test || (( failed_tests++ ))
+  done
+  echo "Failed test suites: ${failed_tests}"
+  if (( failed_tests > 0 )); then
+    exit 1
+  fi
+}
+
+main "$@"
```
Comment (sergiitk) [inline packages/grpc-js-xds/scripts/xds_k8s_lb.sh:166]: Nit: missing new line at the EOF.

File: packages/grpc-js-xds/scripts/xds_k8s_lb.sh [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+main() {
+  local script_dir
+  script_dir="$(dirname "$0")"
+
+  # Source the test driver from the master branch.
+  echo "Sourcing test driver install script from: ${TEST_DRIVER_INSTALL_SCRIPT_URL}"
+  source /dev/stdin <<< "$(curl -s "${TEST_DRIVER_INSTALL_SCRIPT_URL}")"
+
+  activate_gke_cluster GKE_CLUSTER_PSM_SECURITY
+  activate_secondary_gke_cluster GKE_CLUSTER_PSM_SECURITY
```
Comment (sergiitk) [inline packages/grpc-js-xds/scripts/xds_k8s_lb.sh:143]: I think this is in error in the python file, and this should be
```suggestion
  activate_gke_cluster GKE_CLUSTER_PSM_LB
  activate_secondary_gke_cluster GKE_CLUSTER_PSM_LB
```

Similar to cpp: https://github.com/grpc/grpc/blob/master/tools/internal_ci/linux/grpc_xds_k8s_lb.sh

File: packages/grpc-js-xds/interop/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+# Make a grpc-node directory and copy the repo into it.
+WORKDIR /node/src/grpc-node
+COPY . .
+
+WORKDIR /node/src/grpc-node/packages/grpc-js
+RUN npm install
+WORKDIR /node/src/grpc-node/packages/grpc-js-xds
+RUN npm install
+
+ENTRYPOINT [ "node", "/node/src/grpc-node/packages/grpc-js-xds/build/interop/xds-interop-client" ]
```
Comment (sergiitk) [inline packages/grpc-js-xds/interop/Dockerfile:30]: I think the issue was with using `WORKDIR` multiple times. WORKDIR sets the path relative to which the rest of commands are executed. Consider this:

```suggestion
RUN cd packages/grpc-js && npm install
RUN cd packages/grpc-js-xds && npm install

ENTRYPOINT [ "node", "packages/grpc-js-xds/build/interop/xds-interop-client" ]
```

**Review by sergiitk — COMMENTED** [review]
The image size over 1 GB is pretty surprising. I'd recommend investing a bit more time into reducing the size. I'll give it a try and report back.

**Comment by sergiitk** [comment]
@murgatroid99 Take a look at this article on building node images: https://docs.docker.com/language/nodejs/build-images/

AFAIK `RUN npm install --production` and https://docs.docker.com/language/nodejs/build-images/#create-a-dockerignore-file should reduce the size at least partially.

**Comment by murgatroid99** [comment]
I did manage to decrease the Docker image size with a build step. I tried again and it worked, and I'm not sure what I did differently, but I'll take it.

File: packages/grpc-js-xds/interop/Dockerfile (hunk 4 above)
Comment (murgatroid99) [reply packages/grpc-js-xds/interop/Dockerfile]: I am pretty sure that this doesn't matter now that my entrypoint is an absolute path.

File: packages/grpc-js-xds/interop/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+FROM node:16-alpine as build
+
+# Make a grpc-node directory and copy the repo into it.
+WORKDIR /node/src/grpc-node
+COPY . .
+
+WORKDIR /node/src/grpc-node/packages/grpc-js
+RUN npm install
+WORKDIR /node/src/grpc-node/packages/grpc-js-xds
+RUN npm install
```
Comment (sergiitk) [inline packages/grpc-js-xds/interop/Dockerfile:28]: I still recommend against using `WORKDIR` for just just changing a folder. Note that it's also creating a folder if it doesn't exist.

```suggestion
RUN cd packages/grpc-js && npm install
RUN cd packages/grpc-js-xds && npm install
```

**Review by sergiitk — APPROVED** [review]
LGTM. Thanks for the improvement.

File: packages/grpc-js-xds/interop/Dockerfile (hunk 5 above)
Comment (sergiitk) [reply packages/grpc-js-xds/interop/Dockerfile:28]: https://docs.docker.com/engine/reference/builder/#workdir

**Comment by murgatroid99** [comment]
For some reason, that last commit breaks the tests, so I am undoing it.

**Comment by murgatroid99** [comment]
Passing test run against this branch: https://source.cloud.google.com/results/invocations/b8fc66d3-1931-46b6-aec4-40d6f72f93b0/targets

**Review by nicolasnoble — APPROVED** [review]

**Pull request merged**